deepspeed/runtime/zero/mics.py

```diff
@@ -0,0 +1,476 @@
+"""
+"Copyright 2020 The Microsoft DeepSpeed Team.
+Licensed under the MIT license.
+"""
+
+import sys
+from typing import List
+
+import torch
+from deepspeed import comm as dist
+from deepspeed.runtime.config import DeepSpeedConfig
+from deepspeed.runtime.zero.mics_utils import (MiCS_CommGroups,
+                                               create_mics_comm_groups,
+                                               scale_tensors)
+from deepspeed.runtime.zero.parameter_offload import (DeepSpeedZeRoOffload,
+                                                      is_zero_param)
+from deepspeed.runtime.zero.partition_parameters import Init, AllGatherCoalescedHandle, ZeroParamStatus
+from deepspeed.runtime.zero.stage3 import DeepSpeedZeroOptimizer_Stage3
+from deepspeed.utils import instrument_w_nvtx, log_dist
+from torch import Tensor
+from torch.nn import Parameter
+
+
+def has_hierarchical_all_gather_groups(comm_groups: MiCS_CommGroups):
+    result = False
+    if comm_groups.param_intra_node_group is not None and comm_groups.param_inter_node_shard_group is not None:
+        result = True
+    return result
+
+
+class MiCS_AllGatherCoalescedHandle(AllGatherCoalescedHandle):
+    """ This handle assumes that no need to
+    copy data out from a contiguous tensor
+    """
+    def __init__(self,
+                 allgather_handle,
+                 params: List[Parameter],
+                 partitions: List[Tensor],
+                 world_size: int) -> None:
+        super().__init__(allgather_handle, params, partitions, world_size)
+
+    def wait(self) -> None:
+        """
+        """
+
+        if self.complete:
+            return
+
+        # let the all-gather op enqueue
+        instrument_w_nvtx(self.allgather_handle.wait)()
+        if self.complete:
+            return
+
+        for _, param in enumerate(self.params):
+            assert param.ds_status == ZeroParamStatus.INFLIGHT, f"expected param {param.ds_summary()} to be inflight"
+            param.ds_status = ZeroParamStatus.AVAILABLE
+
+        self.complete = True
+
+
+class MiCS_Init(Init):
```

Comment:
I wonder if we could re-use the existing `deepspeed.zero.Init()` context manager? then we could create the mics_comm_groups if mics is enabled?

Reply:
Yes. Do you think it would be better to wrap a class on top of `deepspeed.zero.Init` or directly modify class `Init`.

Reply:
Unless I misunderstand, the only thing that `MiCS_Init` does that zero.Init doesn't is create the `mics_comm_groups` here:

https://github.com/dmlc/DeepSpeed/blob/05afdbd02a687ce96184c221bae023a403f3e8d4/deepspeed/runtime/zero/mics.py#L79-L83

Can we just change zero.Init to add the following:

```python
if _ds_config is not None and _ds_config.mics_shard_size > 0:
    self.mics_comm_groups = create_mics_comm_groups(...)
```

We could add it here:

https://github.com/dmlc/DeepSpeed/blob/05afdbd02a687ce96184c221bae023a403f3e8d4/deepspeed/runtime/zero/partition_parameters.py#L656

Reply:
As @samadejacobs pointed out to me, I missed all the other things that `MiCS_Init` is doing here. I think how you have it currently is fine. Two pending items on this though:

1) can you add a docstring for MiCS_Init?
2) we'll need to add a section to the z3 docs page to describe how to use MiCS: https://github.com/microsoft/DeepSpeed/blob/master/docs/code-docs/source/zero3.rst

Reply:
Do we agree with the current API modification?
Thanks for pointing out the docstring requirement. 
Will work on merging the master branch and the docstring. 

Thanks @samadejacobs, for reviewing the code changes.

Reply:
@zarzen , yes, current API modification is approved. We are ready to merge as soon as two items highlighted above are done, many thanks for your contribution.